client/state/http/index.js

```diff
@@ -67,22 +61,25 @@ export const httpHandler = ( { dispatch }, action ) => {
 		)
 		.join( '&' );
 
-	if ( queryString.length > 0 ) {
-		request.query( queryString );
+	let response, json;
+	try {
+		response = await fetch( queryString.length ? `${ url }?${ queryString }` : url, {
+			method,
+			headers: fetchHeaders,
+			body,
+			credentials: withCredentials ? 'include' : 'same-origin',
+		} );
+		json = await response.json();
+	} catch ( error ) {
+		dispatch( extendAction( onFailure, failureMeta( error ) ) );
+		return;
 	}
 
-	headers.forEach( ( [ headerKey, headerValue ] ) => request.set( headerKey, headerValue ) );
-
-	request.accept( 'application/json' );
-
-	if ( body ) {
-		request.send( body );
+	if ( response.ok ) {
+		dispatch( extendAction( onSuccess, successMeta( { body: json } ) ) );
+	} else {
+		dispatch( extendAction( onFailure, failureMeta( { response: { body: json } } ) ) );
```

Comment:
`response.ok` means a 200 status code, right? Did `superagent` also throw a error (rejected promise) in case of a 400 or 500 response? And what was the shape of the error object.

I'm a bit concerned that the object passed to the failure action is not very well defined. Sometimes it's an `Error` instance, sometimes it's a response object... That makes error handling hard.

Reply:
> `response.ok` means a 200 status code, right?

It's `true` if the response is in the 200-299 range.

> Did `superagent` also throw a error (rejected promise) in case of a 400 or 500 response? And what was the shape of the error object.

Yes, it did. It threw an `Error` object that looks like this when stringified:

```json
{
  "original": null,
  "response": {
    "req": {
      "method": "GET",
      "url": "https://sgom.es/missing",
      "headers": {}
    },
    "xhr": {},
    "text": "SERIALIZED RESPONSE GOES HERE",
    "statusText": "",
    "statusCode": 404,
    "status": 404,
    "statusType": 4,
    "info": false,
    "ok": false,
    "redirect": false,
    "clientError": true,
    "serverError": false,
    "error": {
      "status": 404,
      "method": "GET",
      "url": "https://sgom.es/missing"
    },
    "created": false,
    "accepted": false,
    "noContent": false,
    "badRequest": false,
    "unauthorized": false,
    "notAcceptable": false,
    "forbidden": false,
    "notFound": true,
    "unprocessableEntity": false,
    "headers": {
      "content-type": "text/html; charset=utf-8"
    },
    "header": {
      "content-type": "text/html; charset=utf-8"
    },
    "type": "text/html",
    "charset": "utf-8",
    "links": {},
    "body": null
  },
  "status": 404
}
```

> 
> I'm a bit concerned that the object passed to the failure action is not very well defined. Sometimes it's an `Error` instance, sometimes it's a response object... That makes error handling hard.

I'm not sure if it makes a huge difference to use an `Error` or a plain object in the dispatch, but I'm happy to switch the code around to an `Error`. Do you think I need to implement the whole structure as above, though?

Reply:
> I'm not sure if it makes a huge difference to use an Error or a plain object in the dispatch

It's not about the exact type of the object, but rather that the question: "what is this error object and how do I use it" doesn't have an obvious answer. Or does it?

It should be fixed/clarified at the data layer level, I don't think it can be solved in this PR. Other data layer handlers still use `superagent`, others are based on `wpcom.js`. Unifying the error handling would have to include them all.

Reply:
> It should be fixed/clarified at the data layer level, I don't think it can be solved in this PR. Other data layer handlers still use `superagent`, others are based on `wpcom.js`. Unifying the error handling would have to include them all.

That makes sense, thanks! I'll start working on other bits and pieces once the foundation is established with this PR merged, with the goal of removing `superagent` off of client code altogether.